grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -467,32 +468,39 @@ private void executePrepared(
   }
 
   @NotNull
-  private BiConsumer<ResponseAndTraceId, Throwable> executeTracingQueryIfNeeded(
+  private BiConsumer<Optional<ResponseAndTraceId>, Throwable> executeTracingQueryIfNeeded(
       Connection connection,
       StreamObserver<Response> responseObserver,
       boolean tracingEnabled,
       ConsistencyLevel consistencyLevel) {
     return (responseAndTraceId, t) -> {
       if (t != null) {
         handleException(t, responseObserver);
-      } else if (!tracingEnabled || responseAndTraceId.tracingIdIsEmpty()) {
+      } else if (!responseAndTraceId.isPresent()) {
+        // ignore, it means that there was an exception handled in the previous step
+        return;
```

Comment:
This works, but why even execute this method if something went wrong at the previous step?

More generally I think the design of this class would be improved if we could manage to turn it into a chain of `thenApply / thenCompose` transformations, instead of relying on `handle / whenComplete` (this is probably off-topic for this PR, and better left as a follow-up).

I'm not sure if I'm explaining this clearly so trying to illustrate on a dummy example:

```java
  private static CompletionStage<Integer> someAsyncOperation(int i) {
    ... // whatever
  }

  public static void main(String[] args) {
    CompletionStage<String> stringFuture = ... // whatever

    // current approach:
    CompletionStage<Optional<Integer>> intFuture =
        stringFuture.handle(
            (s, t) -> {
              if (t != null) {
                t.printStackTrace();
                return Optional.empty();
              } else {
                return Optional.of(Integer.parseInt(s));
              }
            });
    intFuture.whenComplete(
        (maybeI, t) -> {
          if (t != null) {
            t.printStackTrace();
          } else if (!maybeI.isPresent()) {
            // ignore because it means error on previous step
          } else {
            someAsyncOperation(maybeI.get())
                .whenComplete(
                    (j, t2) -> {
                      if (t2 != null) {
                        t2.printStackTrace();
                      } else {
                        System.out.printf("The result is %d%n", j);
                      }
                    });
          }
        });

    // proposed approach:
    stringFuture
        .thenApply(Integer::parseInt)
        .thenCompose(EnclosingClass::someAsyncOperation)
        .whenComplete(
            (r, t) -> {
              if (t != null) {
                t.printStackTrace();
              } else {
                System.out.printf("The result is %d%n", r);
              }
            });
  }
```

The advantage of the second approach is that exceptions are propagated naturally and we only need to handle them once at the end.
`whenComplete` is a bit dangerous because if you forget to chain the next action in one of the branches, or if an exception is thrown and uncaught, it can easily leave a request hanging.

Reply:
I agree that's a good idea. I was also thinking that the current approach could be improved.
From the simple experiment with the usage of the `thenCompose` I see one problem:
the signature of method would need to be:
`private Function<Result, CompletionStage<ResponseAndTraceId>> handleQuery(`
However, we need to have the `throwable` in the `handleQuery` to execute the retry logic:
```
 if (t instanceof PreparedQueryNotFoundException) {
          prepareQuery(connection, prepareInfo, tracingEnabled, true)
              .whenComplete(
                  (p, t1) -> {
                    if (t1 != null) {
                      handleException(t1, responseObserver);
                    } else {
                      executePrepared(connection, p, query, responseObserver, prepareInfo);
                    }
                  });
        } else {
          handleException(t, responseObserver);
```
so it may not be so straightforward. But overall, I agree that it may deserve some refactoring. 
@olim7t can you create a separate ticket for this refactoring?

Reply:
> so it may not be so straightforward.

You're right, I also realized that after I posted the comment.

> @olim7t can you create a separate ticket for this refactoring?

Will do.